File: C Programs/2nd sem C programs/quick sort.c
#include<stdio.h>
int main()
{
	void quick_sort(int [],int );
	int partition(int [],int ,int );
	int a[100],n,i,ch;
	printf("How many values you want to insert: ");
	scanf("%d",&n);
	for(i=0;i<n;i++)
	{
		printf("enter a[%d]= ",i+1);
		scanf("%d",&a[i]);
	}
	quick_sort(a,n);
	printf("\nThe array after quick sort:\n");
	for(i=0;i<n;i++)
	{
		printf("%d ",a[i]);
	}
	return 1;	
}
int partition(int a[],int left,int right)
{
	int pos,t;
	pos=left;
	while(left<=right)
	{
		while(a[pos]<=a[right]&&pos!=right)
	    	right--;
	    if(pos==right)
			return(pos);
		else
		{
			t=a[pos];
			a[pos]=a[right];
			a[right]=t;
			pos=right;
		}
		while(a[left]<=a[pos]&&left!=pos)
			left++;
		if(pos==left)
			return(pos);
		else
		{
			t=a[pos];
			a[pos]=a[left];
			a[left]=t;
			pos=left;
		}				
	}
}
void quick_sort(int a[],int n)
{
	int lower[50],upper[50],pos,beg,end,top=-1;
	if(n<=1)
		return;
	top=top+1;
	lower[top]=0;
	upper[top]=n-1;
	while(top!=-1)
	{
		beg=lower[top];
		end=upper[top];
		top=top-1;
		pos=partition(a,beg,end);
		if(beg+1<pos)
		{
			top=top+1;
			lower[top]=beg;
			upper[top]=pos-1;
		}
		if(end>pos+1)
		{
			top=top+1;
			lower[top]=pos+1;
			upper[top]=end;
		}
	}	
}

File: Java Program/Sorting/Quick_sort.java
import java.util.*;
import java.io.*;

class Quick_sort{
    
    int partition(int a[], int left, int right)
    {
        int pos, t;
        
        pos=left;
        while(left<=right)
        {
            while(a[pos]<=a[right]&&pos!=right)
            {
                right--;
            }
            if(pos==right)
                return (pos);
            else
            {
                t=a[pos];
                a[pos]=a[right];
                a[right]=t;
                pos=right;
            }
            while(a[left]<=a[pos]&&pos!=left)
            {
                left++;
            }
            if(pos==left)
                return (pos);
            else
            {
                t=a[pos];
                a[pos]=a[left];
                a[left]=t;
                pos=left;
            }
        }
        return (pos);
    }
    
    void quick_sort_rc(int a[], int l, int r)
    {
        int i;
        if(l<r)
        {
            i=partition(a, l, r);
            quick_sort_rc(a,l,i-1);
            quick_sort_rc(a, i+1, r);
        }
    }
    
    void display(int a[], int n)
    {
        int i;
        System.out.println("The array after quick sort is : \n");

        for(i=0;i<n;i++)
        {
            System.out.print(a[i]+" ");
        }
        System.out.println("\n");
    }
    
    public static void main(String args[])
    {
        int a[], n, i;
        
        Quick_sort obj = new Quick_sort();
        
        Scanner sc= new Scanner(System.in);
        
        System.out.print("\nEnter the number of element: ");
        
        n=sc.nextInt();
        
        a=new int[n];
        
        System.out.println();
        for(i=0;i<n;i++)
        {
            System.out.print("Enter a["+i+"] : ");
            a[i]=sc.nextInt();
            System.out.println();
        }
        
        //Quick Sort
        
        obj.quick_sort_rc(a,0,n-1);
        obj.display(a,n);
    
    }

}
